engine/kv_engine_cleaner.cpp

```diff
@@ -0,0 +1,466 @@
+        kvdk_assert(skiplist_node->record->GetRecordType() == SortedElemDelete,
+                    "Should be sorted elem delete type");
+        pending_free_entries->emplace_back(
+            purgeSortedRecord(skiplist_node, skiplist_node->record));
+        break;
+      }
+      case PointerType::DLRecord: {
+        DLRecord* dl_record = static_cast<DLRecord*>(record_pair.first);
+        kvdk_assert(dl_record->GetRecordType() == SortedElemDelete,
+                    "Should be sorted elem delete type");
+        pending_free_entries->emplace_back(
+            purgeSortedRecord(nullptr, dl_record));
+        break;
+      }
+      default:
+        std::abort();
+    }
+  }
+}
+
+struct PendingPurgeOldRecords {
+  std::vector<PMemOffsetType> records;
+  TimeStampType release_time;
+};
+
+struct PendingPurgeOutDatedRecords {
+  std::vector<std::pair<void*, PointerType>> outdated_records;
+  TimeStampType release_time;
+};
+
+void KVEngine::CleanOutDated(size_t start_slot_idx, size_t end_slot_idx) {
+  std::deque<PendingFreeSpaceEntries> pending_free_spaces;
+  std::deque<std::pair<TimeStampType, List*>> outdated_lists;
+  std::deque<std::pair<TimeStampType, HashList*>> outdated_hash_lists;
+  std::deque<std::pair<TimeStampType, Skiplist*>> outdated_skip_lists;
+  std::deque<PendingPurgeOldRecords> pending_purge_strings;
+  std::deque<PendingPurgeOldRecords> pending_purge_dls;
+  std::deque<PendingPurgeOutDatedRecords> pending_purge_outdated;
+
+  while (!bg_work_signals_.terminating) {
+    size_t total_num = 0;
+    size_t need_purge_num = 0;
+    size_t slot_num = 0;
+
+    version_controller_.UpdatedOldestSnapshot();
+
+    std::vector<PMemOffsetType> old_string_records;
+    std::vector<PMemOffsetType> old_dl_records;
+    std::vector<std::pair<void*, PointerType>> outdated_records;
+
+    // Iterate hash table
+    auto hashtable_iter =
+        hash_table_->GetIterator(start_slot_idx, end_slot_idx);
+    while (hashtable_iter.Valid()) {
+      auto min_snapshot_ts = version_controller_.OldestSnapshotTS();
+      auto now = TimeUtils::millisecond_time();
+
+      {  // Slot lock section
+        auto slot_lock(hashtable_iter.AcquireSlotLock());
+        auto slot_iter = hashtable_iter.Slot();
+        while (slot_iter.Valid()) {
+          if (!slot_iter->Empty()) {
+            total_num++;
+            switch (slot_iter->GetIndexType()) {
+              case PointerType::StringRecord: {
+                auto string_record = slot_iter->GetIndex().string_record;
+                if ((string_record->GetRecordType() ==
+                         RecordType::StringDeleteRecord ||
+                     string_record->GetExpireTime() <= now) &&
+                    string_record->GetTimestamp() < min_snapshot_ts) {
+                  need_purge_num++;
+                  outdated_records.emplace_back(
+                      std::make_pair(string_record, slot_iter->GetIndexType()));
```

Comment:
This queue holds std::pair<void*, PointerType>, while the other queue holds PMemOffsetType?
Similar things should have similar or the same type.